File: app/src/main/java/djordjeh/architecture/mvp/data/TaskRepository.java
package djordjeh.architecture.mvp.data;

import android.support.annotation.NonNull;

import java.util.List;

import djordjeh.architecture.mvp.data.model.Task;
import djordjeh.architecture.mvp.data.source.TaskDataSource;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;

public class TaskRepository implements TaskDataSource {

    private final TaskDataSource localDataSource;
    private final TaskDataSource remoteDataSource;

    TaskRepository(TaskDataSource localDataSource, TaskDataSource remoteDataSource) {
        this.localDataSource = localDataSource;
        this.remoteDataSource = remoteDataSource;
    }

    @Override
    public Observable<List<Task>> tasks(boolean forceUpdate) {
        if (forceUpdate) return remoteDataSource.tasks(true)
                .flatMap(tasks -> Observable.fromIterable(tasks)
                        .doOnNext(localDataSource::save)
                        .toList()
                        .toObservable());
        else return localDataSource.tasks(false);
    }

    @Override
    public Maybe<Task> task(long taskId) {
        return Maybe.concat(localDataSource.task(taskId), remoteDataSource.task(taskId)).firstElement();
    }

    @Override
    public Single<Task> save(@NonNull Task task) {
        return remoteDataSource.save(task)
                .flatMap(localDataSource::save);
    }

    @Override
    public Single<Task> delete(@NonNull Task task) {
        return remoteDataSource.delete(task)
                .flatMap(localDataSource::delete);
    }
}

File: app/src/main/java/djordjeh/architecture/mvp/data/TaskRepository.kt
package djordjeh.architecture.mvp.data

import djordjeh.architecture.mvp.data.model.Task
import djordjeh.architecture.mvp.data.source.TaskDataSource
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single

class TaskRepository internal constructor(private val localDataSource: TaskDataSource, private val remoteDataSource: TaskDataSource) : TaskDataSource {

    override fun tasks(forceUpdate: Boolean): Observable<List<Task>> {
        return if (forceUpdate)
            remoteDataSource.tasks(true)
                    .flatMap { tasks ->
                        Observable.fromIterable(tasks)
                                .doOnNext { task: Task -> localDataSource.save(task) }
                                .toList()
                                .toObservable()
                    }
        else
            localDataSource.tasks(false)
    }

    override fun task(taskId: Long): Maybe<Task> {
        return Maybe.concat(localDataSource.task(taskId), remoteDataSource.task(taskId)).firstElement()
    }

    override fun save(task: Task): Single<Task> {
        return remoteDataSource.save(task)
                .flatMap { localDataSource.save(it) }
    }

    override fun delete(task: Task): Single<Task> {
        return remoteDataSource.delete(task)
                .flatMap { localDataSource.delete(it) }
    }
}
